lib/handler/RetryHandler.js

```diff
@@ -0,0 +1,347 @@
+const assert = require('node:assert')
+
+const { kRetryHandlerDefaultRetry } = require('../core/symbols')
+const { RequestRetryError } = require('../core/errors')
+const {
+  isDisturbed,
+  parseHeaders,
+  parseRangeHeader,
+  safeHTTPMethods
+} = require('../core/util')
+
+function calculateRetryAfterHeader (retryAfter) {
+  const current = Date.now()
+  const diff = new Date(retryAfter).getTime() - current
+
+  return diff
+}
+
+class RetryHandler {
+  constructor (opts, handlers, retryOpts) {
+    const {
+      // Retry scoped
+      retry: retryFn,
+      max,
+      maxTimeout,
+      minTimeout,
+      timeoutFactor,
+      // Response scoped
+      methods,
+      idempotent,
+      codes,
+      retryAfter,
+      status: statusCodes
+    } = retryOpts ?? {}
+
+    this.dispatch = handlers.dispatch
+    this.handler = handlers.handler
+    this.opts = opts
+    this.abort = null
+    this.aborted = false
+    this.retryOpts = {
+      retry: retryFn ?? RetryHandler[kRetryHandlerDefaultRetry],
+      retryAfter: retryAfter ?? true,
+      maxTimeout: maxTimeout ?? 30 * 1000, // 30s,
+      timeout: minTimeout ?? 500, // .5s
+      timeoutFactor: timeoutFactor ?? 2,
+      max: max ?? 5,
+      // What errors we should retry
+      methods: methods ?? ['GET', 'HEAD', 'OPTIONS', 'PUT', 'DELETE', 'TRACE'],
+      // States weather or not retry on idempotent methods
+      idempotent: idempotent ?? false,
+      // Indicates which errors to retry
+      status: statusCodes ?? [500, 502, 503, 504, 429],
+      // List of errors to retry
+      codes: codes ?? [
+        'ECONNRESET',
+        'ECONNREFUSED',
+        'ENOTFOUND',
+        'ENETDOWN',
+        'ENETUNREACH',
+        'EHOSTDOWN',
+        'EHOSTUNREACH',
+        'EPIPE'
+      ]
+    }
+
+    // TODO: hidde behind a symbol
+    this.count = 0
+    this.timeout = null
+    this.retryAfter = null
+    this.head = 0
+    this.tail = null
```

Comment:
start and end is a bit clearer? Took me a bit to figure out what these were for. Maps better to existing vocabulary in regards to ranges.

Reply:
SGTM
I don't recall exactly why I went in that direction, I kind of recall because wanted to handle those two differently; but yeah your suggestion goes better 👍